### ISSUE TITLE
Direct rules do not override package rename rules

### ISSUE THREAD

**Issue by jhanders34** [eclipse/transformer#190]
When moving from JAXB 2.x to XMLBinding 3.0 there is a situation where the package renames are:

com.sun.xml.bind changes to org.glassfish.jaxb.runtime

The property anme com.sun.xml.bind.ObjectFactory though changes to org.glassfish.jaxb.core.ObjectFactory. 

Looking at the code it takes package rename as precedence instead of direct for Strings.

**Comment by bjhargrave** [comment]
This is more of a type rename rule rather than a string rename rule.

**Comment by jhanders34** [comment]
To give an idea, the rename rule will be:
com.sun.xml.bind=org.glassfish.jaxb.runtime

And the direct rule will be:
com.sun.xml.bind.ObjectFactory=org.glassfish.jaxb.core.ObjectFactory

**Comment by bjhargrave** [comment]
> Tom is going to change it to take precedence when transforming a String.

I am not sure this is a good idea and needs lots of thought and testing first.

Package renames are looking at class references while string changes are just string constants without respect to whether the string constant is backing up a class reference. It would be much safer to introduce the noting of type renaming (`com.sun.xml.bind.ObjectFactory=org.glassfish.jaxb.core.ObjectFactory`) which happens before package renaming. Type renaming would rename class references rather than just modify string constants.

**Comment by bjhargrave** [comment]
> The point was to ONLY do this when doing a String constant.

So a class reference to `com.sun.xml.bind.ObjectFactory` should not be changed to a class reference to `org.glassfish.jaxb.core.ObjectFactory`? If you change just the string constant, this wont affect the class reference.

**Comment by bjhargrave** [comment]
I think I was confused by the fact that package renaming rules contributed to string constant changes at all. I had assumed that package renaming rules applied only to class constants and string changing rules applied only to string constants. Instead, we have a situation where package renaming rules can apply to both class constants and string constants and string renaming rules can also apply to string constants so we need to diddle with ordering of rule application to string constants.

It would be much cleaner to have a set of rules which apply to class constants only and a set of rules which apply to string constants only.

But given the rationale above, the suggested ordering for modifying string constants appears to make sense.

**Comment by tbitonti** [comment]
Dredging the history of the implementation, the contribution arose so to transform debugging information, which has a lot of embedded class names.
Somewhere we added the case of replacing simple dotted and slashified class names, which helps us with Class.forName and other direct class name references.
On top of those two cases, direct string updates were added, and were probably simply added last in the sequence of updates.
And then on top of that we added per-class string updates, which seem also to have been added at the end.

The core problem seems to be the collision of direct string updates (of either type) with the package rename type updates (again, of either type).  I think its fair to ask if we should be performing the package rename type updates to UTF8 constants, or perhaps should have that be enabled by a switch.  Likewise, if both types of updates are being performed, what is the correct ordering?  Should that also be configurable?  (I'm not sure having a lot of configuration options of this sort is a good idea.)
